### ISSUE TITLE
Check and document that bottom diffs are written, while param diffs are accumulated

### ISSUE THREAD

**Issue by erogol** [BVLC/caffe#3248]
I printed the initial values, just before of the diff computation in SoftmaxWithLoss layer and the values are not zero which is contrary to my expectation. This is a small buggy thing or I am missing something.

These are the valuesfor batch_size = 5;

````
I1027 14:09:53.631011 19786 net.cpp:283] Network initialization done.
I1027 14:09:53.633285 19786 solver.cpp:59] Solver scaffolding done.
I1027 14:09:53.634878 19786 caffe.cpp:212] Starting Optimization
I1027 14:09:53.634903 19786 solver.cpp:287] Solving inception_bn
I1027 14:09:53.634912 19786 solver.cpp:288] Learning Rate Policy: step
I1027 14:10:05.057032 19786 softmax_loss_layer.cpp:120] 0.00132518
I1027 14:10:05.057107 19786 softmax_loss_layer.cpp:120] 0.000371205
I1027 14:10:05.057116 19786 softmax_loss_layer.cpp:120] 0.000937449
I1027 14:10:05.057121 19786 softmax_loss_layer.cpp:120] 0.00039356
I1027 14:10:05.057126 19786 softmax_loss_layer.cpp:120] 0.000761485
I1027 14:13:14.637547 19786 solver.cpp:236] Iteration 0, loss = 7.30357
I1027 14:13:14.637847 19786 solver.cpp:252]     Train net output #0: logprob = 7.30357 (* 1 = 7.30357 loss)
I1027 14:13:14.637872 19786 sgd_solver.cpp:106] Iteration 0, lr = 0.01
````

I add this line to print values;

````
....
LOG(INFO) <<  bottom_diff[i * dim + label_value * inner_num_ + j];
bottom_diff[i * dim + label_value * inner_num_ + j] -= 1 ;
....
````

**Comment by seanbell** [comment]
Yes, the current interface expects that the layer _overwrite_ the diff for the bottom.  Therefore, there is no need to zero them.  For parameters, the diff is zeroed out and the layer _accumulates_.

This is a little confusing and also currently not tested by the gradient checker.

**Comment by longjon** [comment]
Closing in favor of #3036.
